sdks/java/extensions/sql/zetasql/src/test/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSqlJavaUdfTest.java

```diff
@@ -0,0 +1,203 @@
+                + "SELECT custom_agg(f_int_1) from aggregate_test_table",
+            jarPath);
+    ZetaSQLQueryPlanner zetaSQLQueryPlanner = new ZetaSQLQueryPlanner(config);
+    BeamRelNode beamRelNode = zetaSQLQueryPlanner.convertToBeamRel(sql);
+    PCollection<Row> stream = BeamSqlRelUtils.toPCollection(pipeline, beamRelNode);
+
+    Schema singleField = Schema.builder().addInt64Field("field1").build();
+
+    PAssert.that(stream).containsInAnyOrder(Row.withSchema(singleField).addValue(0L).build());
+    pipeline.run().waitUntilFinish(Duration.standardMinutes(PIPELINE_EXECUTION_WAITTIME_MINUTES));
+  }
+
+  @Test
+  public void testUnregisteredFunction() {
+    assumeNotNull(jarPath);
+    String sql =
+        String.format(
+            "CREATE FUNCTION notRegistered() RETURNS STRING LANGUAGE java OPTIONS (path='%s'); "
+                + "SELECT notRegistered();",
+            jarPath);
+    ZetaSQLQueryPlanner zetaSQLQueryPlanner = new ZetaSQLQueryPlanner(config);
+    thrown.expect(RuntimeException.class);
+    thrown.expectMessage(
+        String.format("No implementation of scalar function notRegistered found in %s.", jarPath));
+    zetaSQLQueryPlanner.convertToBeamRel(sql);
+  }
+
+  @Test
+  public void testJarContainsNoUdfProviders() {
+    assumeNotNull(jarPath);
+    assumeNotNull(emptyJarPath);
+    String sql =
+        String.format(
+            // Load an inhabited jar first so we can make sure jars load UdfProviders in isolation
+            // from other jars.
+            "CREATE FUNCTION foo() RETURNS STRING LANGUAGE java OPTIONS (path='%s'); "
+                + "CREATE FUNCTION bar() RETURNS STRING LANGUAGE java OPTIONS (path='%s'); "
+                + "SELECT bar();",
+            jarPath, emptyJarPath);
+    ZetaSQLQueryPlanner zetaSQLQueryPlanner = new ZetaSQLQueryPlanner(config);
+    thrown.expect(ProviderNotFoundException.class);
+    thrown.expectMessage(String.format("No UdfProvider implementation found in %s.", emptyJarPath));
+    zetaSQLQueryPlanner.convertToBeamRel(sql);
+  }
+
+  @Test
+  public void testJavaUdfNoJarProvided() {
+    String sql = "CREATE FUNCTION foo() RETURNS STRING LANGUAGE java; SELECT foo();";
+    ZetaSQLQueryPlanner zetaSQLQueryPlanner = new ZetaSQLQueryPlanner(config);
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("No jar was provided to define function foo.");
+    zetaSQLQueryPlanner.convertToBeamRel(sql);
+  }
+
+  @Test
+  public void testPathOptionNotString() {
+    String sql =
+        "CREATE FUNCTION foo() RETURNS STRING LANGUAGE java OPTIONS (path=23); SELECT foo();";
+    ZetaSQLQueryPlanner zetaSQLQueryPlanner = new ZetaSQLQueryPlanner(config);
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Option 'path' has type TYPE_INT64 (expected TYPE_STRING).");
+    zetaSQLQueryPlanner.convertToBeamRel(sql);
+  }
```

Comment:
Can you add a case to verify `NULL` handling? E.g. pass a null value to `increment`? It is ok if it does not work, but it will be good to see expected behavior and then document it.

Reply:
I added some tests. I discovered a gap in type checking, since the current implementation does not strictly enforce that the SQL function signature must match the Java signature. I will fix that.

Reply:
That makes sense. Thanks!